server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -299,6 +300,47 @@ public Metadata withIncrementedVersion() {
         );
     }
 
+    public Metadata withLifecycleState(final Index index, final LifecycleExecutionState lifecycleState) {
+        Objects.requireNonNull(index, "index must not be null");
+        Objects.requireNonNull(lifecycleState, "lifecycleState must not be null");
+
+        // build a new index metadata with the version incremented and the new lifecycle state
+        IndexMetadata indexMetadata = getIndexSafe(index);
+        IndexMetadata.Builder indexMetadataBuilder = IndexMetadata.builder(indexMetadata);
+        indexMetadataBuilder.version(indexMetadataBuilder.version() + 1);
+        indexMetadataBuilder.putCustom(ILM_CUSTOM_METADATA_KEY, lifecycleState.asMap());
+
+        // drop it into the indices
+        final ImmutableOpenMap.Builder<String, IndexMetadata> builder = ImmutableOpenMap.builder(indices);
+        builder.put(index.getName(), indexMetadataBuilder.build());
+
+        return new Metadata(
```

Comment:
I think it might be worth adding a comment about why this doesn't use `Metadata.builder(this).etc(...)` but directly constructs the Metadata, so that it doesn't get accidentally undone in the future. What do you think?

Reply:
I handled this in ea9c861b31fbd129336acf1833b8455deed7336a, too.